tests/trainer/test_trainer.py

```diff
@@ -1937,3 +1937,32 @@ def test_exception_when_lightning_module_is_not_set_on_trainer():
         trainer.test()
     with pytest.raises(MisconfigurationException, match=r"`model` must be provided.*predict"):
         trainer.predict()
+
+
+@RunIf(min_gpus=2, special=True)
+def test_ddp_terminate_when_deadlock_is_detected(tmpdir):
+
+    class CustomException(Exception):
+        pass
+
+    class TestModel(BoringModel):
+
+        def training_step(self, batch, batch_idx):
+            if batch_idx == 1 and self.trainer.is_global_zero:
+                raise CustomException
```

Comment:
@tchaton how is it possible this test works as expected. 
from the code above we can see you wait 3 secs but in this time the other process can easily finish training and should run into an assertion error, because below you test
`with pytest.raises(DeadlockDetectedException, match="CustomException")`
but rank 1 should not raise.

Reply:
is it because process 1 is running into a barrier, while process 0 is sitting in the exception handling?

Reply:
Hey @awaelchli,

Rank 0 is raising an exception and this is going to create a deadlock.

Therefore rank 0 is outside of self.fit_loop.run() and rank 1 on sync.
Rank 0 is going to save a file, wait 3 sec and list the directory. There will be only 1 file as process 1 is blocked in the loop.
Therefore it will kill rank 1 and raise a DeadLockException.